### PR TITLE
RNN 循环神经网络系列 1：基本 RNN 与 CHAR-RNN

### LINKED ISSUE #2247: RNN 循环神经网络系列 1：基本 RNN 与 CHAR-RNN
* 原文链接：[RECURRENT NEURAL NETWORKS (RNN) – PART 1: BASIC RNN / CHAR-RNN](https://theneuralperspective.com/2016/10/04/05-recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md)
  * PR 地址：#2277
  * 文章分类：**机器学习**
  ----
  * 翻译时间：4 天
  * 校对时间：1 天
  * 翻译积分：4 分
  * 校对积分：1 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by changkun**
译文翻译完成，resolve #2247

### CODE REVIEW

**Comment by CACppuccino** [comment]
校对认领

**Comment by fanyijihua** [comment]
@CACppuccino 好的呢 :beer:

File: TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md [hunk 1]
```diff
@@ -370,28 +372,28 @@ class model(object):
...
+注意到 `self.initial_state` 由 `stacked_cell.zero_state` 初始化，我们只需要指定的 `batch_size` 就够了。对于这里的 `NUM_BATCHES` 请查看前面的张量形状一节中的说明。有一种替代方法可以不包含初始状态，`dynamic_rnn()` 会自行处理，我们所需要做的就是指定数据类型（即`dtype = tf.float32` 等)。可惜我们并不能这样做，因为我们要把序列的 `final_state` 作为了下一个序列的 `initial_state` 。你可能还会注意到，尽管 `self.initial_state` 不是占位符，我们还是把前一次 `final_state` 传给了新的 `initial_state`。当然，我们可以通过重载 `step()` 里的 `self.initial_state` 来输入自己的初始值。不管怎样，一旦用到 `input_feeds` ，我们就需要计算 `output_feed`，而如果没有用到，那么就会跳回使用重载之前的值（也就是 `stacked_cell.zero_state`）。
```
Comment (CACppuccino) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: [我们可以通过重载 `step()` ] =>[我们可以通过重新定义 `step()` ]

**Comment by TobiasLee** [comment]
@fanyijihua  校对认领

**Comment by fanyijihua** [comment]
@TobiasLee 妥妥哒 :beers:

File: TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md [hunk 2]
```diff
@@ -3,123 +3,125 @@
...
+    # 但我们每次只想输入 seq_len 个 token
     feed_size = FLAGS.BATCH_SIZE // FLAGS.SEQ_LEN
     for i in range(feed_size):
         X = data_X[:, i * FLAGS.SEQ_LEN:(i+1) * FLAGS.SEQ_LEN]
         y = data_y[:, i * FLAGS.SEQ_LEN:(i+1) * FLAGS.SEQ_LEN]
         yield (X, y)
 ```
 
-Below is the code that uses all of our weights. We have an rnn_cell that takes in the input and the state from the previous cell in order to generate the rnn output which is also the next cell’s input state. The next function, rnn_logits, converts our rnn output using weights to generate logits to be used for probability determination via softmax.
+下面是使用我们自己的权重的代码。`rnn_cell`函数用来接收来自前一个单元的输入和状态，从而生成 RNN 的输出，同时也是下一个单元的输入状态。下一个函数 `rnn_logits` 使用权重将我们的 RNN 输出进行转换，从而通过 softmax 生成 logit 概率并用于分类。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: `rnn_cell`函数用来接收来自前一个单元的输入和状态 =>
`rnn_cell` 函数用来接收来自前一个单元的输入和状态
增加空格

Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: 从而通过 softmax 生成 logit 概率并用于分类 =>
从而通过 softmax 生成 logits 概率并用于分类。

File: TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md [hunk 3]
```diff
@@ -169,15 +171,15 @@ class model(object):
...
+        # 创建 RNN
         state = self.initial_state
         rnn_outputs = []
         for rnn_input in rnn_inputs:
             state = rnn_cell(FLAGS, rnn_input, state)
             rnn_outputs.append(state)
         self.final_state = rnn_outputs[-1]
 
-        # Logits and predictions
+        # Logit 概率及预测
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: Logit 概率及预测 =>
Logits 概率及预测
译者的 Logits 都翻译成 Logit，我认为 Logits 更好，译者是认为强调单个样本的输入吗？我的理由是在大部分文章以及教程中都是 Logits，以及 Tensorflow 的接口也是使用 Logits，翻译成 Logits 保持阅读的一致性可能较好？
望再斟酌一下。

File: TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md [hunk 4]
```diff
@@ -282,30 +284,30 @@ def rnn_cell(FLAGS):
...
 
-    # Each state as one cell
+    # 每个状态作为单个 cell
     stacked_cell = tf.nn.rnn_cell.MultiRNNCell([single_cell] * FLAGS.NUM_LAYERS)
 
     return stacked_cell
 ```
 
-The code above is about creating our specific rnn architecture. We can choose from many different rnn cell types but here you can see three of most common (basic, GRU, and LSTM). We create each cell with a certain number of hidden units. We can then add a dropout layer after cell layer for regularization. Finally, we can make the stacked rnn architecture by replicating the single_cell. Note the **state_is_tuple=True** condition added to single_cell and stacked_cell. This ensures that we get a tuple return that contains the states after each input in a given sequence. The above statement will be true if using an LSTM unit, otherwise, please disregard.
+上面的代码创建的是我们特定的 RNN 结构。我们可以从许多不同的 RNN 单元类型中进行选择，但是在这里你可以看到三个最常见的类型（基本、GRU 和 LSTM）。我们用一定数量的隐藏单元来创建每个 RNN 单元。然后，我们可以在每个单元层之后之后添加一个 Dropout 层来进行正则化处理。最后，我们可以通过复制 `single_cell` 来实现堆叠的 RNN 结构。注意，`state_is_tuple=True` 条件被附加到了 `single_cell` 和 `stacked_cell` 里。这保证了在给定序列的每个输入之后返回一个包含状态的元组。如果使用 LSTM 单元，上述语句为真；否则无视。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: 但是在这里你可以看到三个最常见的类型（基本、GRU 和 LSTM） =>
但是在这里你可以看到三个最常见的类型（BasicRNN、GRU 和 LSTM）

File: TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md [hunk 5]
```diff
@@ -370,28 +372,28 @@ class model(object):
...
+        # 保存模型的组件
         self.global_step = tf.Variable(0, trainable=False)
         self.saver = tf.train.Saver(tf.all_variables())
 ```
 
-Also notice that we don’t manually do one-hot encoding on our input tokens before embeddings them. This is because **tf.nn.embedding_lookup** in rnn_inputs function above does this automatically for us.
+还要注意的是，我们不会手动的在嵌入之前对输入 token 进行独热编码，这是因为`rnn_inputs` 函数里的  `tf.nn.embedding_lookup` 会自动帮我们完成。
 
-For generating the outputs, we use **tf.nn.dynamic_rnn** where the outputs will be the output for each input and the returning state is a tuple containing the last state for each input batch. Finally, we reshape the outputs to shape so we can get the logits and compare to targets.
+为了生成输出，我们使用了 `tf.nn.dynamic_rnn` ，其输出结果为每个输入的输出以及返回状态（即包含上一次每个输入批次的状态的元组）。最后，我们将输出进行了 reshape ，从而得到 logit 概率并用于与targets 进行比较。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-1-basic-rnn-char-rnn.md]: 从而得到 logit 概率并用于与targets 进行比较。 =>
从而得到 logit 概率并用于与 targets 进行比较。

**Comment by changkun** [comment]
@TobiasLee 非常感谢校对和指出的相关意见！关于 logits 的问题你说得很对，我本身在接受这个概念的时候对于 logit 的理解是一个单一的概率（单数名词），但实际在网络的计算中涉及很多 logit 概率（名词复数），因此实际实现中写作 logits。我期初翻译的时候并不是特别清楚中文文章中该有的姿势，但你说得很对，为了考虑中文的习惯和读者的方便起见还是改为 logits 为好。

@leviding 本文校对完毕。

**Comment by leviding** [comment]
@changkun 发布文章反馈链接啦

**Comment by changkun** [comment]
@leviding sorry... 发完文章后就搞忘了.. https://juejin.im/post/59f0c5b0f265da43085d3e94
